File: policy-engine/src/polisyos/common/markdown.py
"""Internal, source-preserving syntax helpers for Markdown table consumers."""

from __future__ import annotations

import re


def split_markdown_table_row(text: str) -> list[str]:
    """Split a row at unescaped pipes outside matched inline-code spans.

    Backtick spans use equal-length maximal runs; unmatched runs are literal.
    Backslashes escape the next character outside code and are literal inside
    code. Optional outer delimiter pipes and padding outside those delimiters
    are omitted. Empty cells and every character inside a cell are preserved,
    including whitespace, escapes and backticks. This does not render Markdown
    or validate column roles, row shape, provenance or authority.

    Args:
        text: One Markdown table row, with or without outer delimiter pipes.

    Returns:
        Source slices for the cells, without unescaping or trimming their text.
    """
    runs = list(re.finditer(r"`+", text))
    next_end_by_width: dict[int, int] = {}
    code_runs: dict[int, tuple[int, int | None]] = {}
    for run in reversed(runs):
        width = run.end() - run.start()
        code_runs[run.start()] = (run.end(), next_end_by_width.get(width))
        next_end_by_width[width] = run.end()

    cells: list[str] = []
    start = 0
    index = 0
    while index < len(text):
        character = text[index]
        if character == "\\":
            index += 2
            continue
        if index in code_runs:
            run_end, closing_end = code_runs[index]
            index = closing_end if closing_end is not None else run_end
            continue
        if character == "|":
            cells.append(text[start:index])
            start = index + 1
        index += 1
    cells.append(text[start:])

    if len(cells) > 1 and not cells[0].strip():
        cells.pop(0)
    if len(cells) > 1 and not cells[-1].strip():
        cells.pop()
    return cells

File: policy-engine/apps/runtime-dashboard/src/features/trust/domain/markdown.ts
/** Split source cells without unescaping, trimming or admitting their semantics. */
export function splitMarkdownTableRow(text: string): string[] {
  // Equal-length maximal backtick runs delimit code; unmatched runs are literal.
  const runs = [...text.matchAll(/`+/gu)];
  const nextEndByWidth = new Map<number, number>();
  const codeRuns = new Map<number, { end: number; closingEnd?: number }>();
  for (const run of runs.reverse()) {
    const start = run.index;
    const width = run[0].length;
    const end = start + width;
    codeRuns.set(start, { end, closingEnd: nextEndByWidth.get(width) });
    nextEndByWidth.set(width, end);
  }

  const cells: string[] = [];
  let start = 0;
  let index = 0;
  while (index < text.length) {
    const character = text[index];
    if (character === "\\") {
      index += 2;
      continue;
    }
    const codeRun = codeRuns.get(index);
    if (codeRun) {
      index = codeRun.closingEnd ?? codeRun.end;
      continue;
    }
    if (character === "|") {
      cells.push(text.slice(start, index));
      start = index + 1;
    }
    index += 1;
  }
  cells.push(text.slice(start));

  // Discard only optional outer delimiters and padding outside those delimiters.
  if (cells.length > 1 && cells[0]!.trim() === "") cells.shift();
  if (cells.length > 1 && cells.at(-1)!.trim() === "") cells.pop();
  return cells;
}
